skmorf/tree/_unsup_criterion.pyx

```diff
@@ -100,26 +151,50 @@ cdef class TwoMeans(Criterion):
         samples : array-like, dtype=SIZE_t
             A mask on the samples, showing which ones we want to use
         """
-        # need to implement
-        pass
+        self.X = X
+        self.sample_weight = sample_weight
+        self.weighted_n_samples = weighted_n_samples
+        self.sample_indices = sample_indices
+
+        return 0
         
     cdef int reset(self) nogil except -1:
         """Reset the criterion at pos=start.
 
         Returns -1 in case of failure to allocate memory (and raise MemoryError)
         or 0 otherwise.
         """
-        # need to implement
-        pass
+        self.pos = self.start
+
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = self.weighted_n_node_samples
+        cdef SIZE_t k
+
+        # TODO: Again UnsupervisedCriterion has no notion of "classes", so irrelevant here.
+        # However, you need to reset sum_left and sum_right right?
+        for k in range(self.n_outputs):
+            memset(&self.sum_left[k], 0, sizeof(double))
+            memcpy(&self.sum_right[k], &self.sum_total[k], sizeof(double))
```

Comment:
```suggestion
        self.sum_left = 0.0
        self.sum_right = self.sum_total
```

You should be able to fill in the other functions, reverse_reset, with the correct implementations. If not, then please write down questions for what you don't understand.